### ISSUE TITLE
Pause/Resume functionality inconsistent

### ISSUE THREAD

**Issue by TheJoshW** [KevinOConnor/klipper#3925]
Hi,
Pause/Resume in Klipper is somewhat complicated and inconsistent.  This is highlighted when using a filament runout plugin.

In my use case, I use Repetier-Server to stream GCODE to the printer.  On runout, the filament runout plugin uses the pause_resume plugin to pause the print, save the GCODE state and "parks" the print head (by specifying runour_gcode).  I then go to the printer and change the filament and use Repetier-Server resume button (or the printer LCD octoprint->resume option) to restart the print.

This fails because neither of these options (Repetier-Server resume button or printer LCD resume) restores the GCODE state on resume, they simply send action:resumed or action:resume respectively to the host to resume streaming GCODE to the printer (which promptly resumes printing in the air as the park lifts the print head before moving it away from the print).

As an interim (and I would be keen for a better solution if possible), I have added GCODE macros to Repetier-Server to be called after pause (PAUSE macro) and before resume (RESUME macro) (very similar to how OctoPrint does it).  I also overrode menu.cfg (in printer.cfg) [menu __main __octoprint __pause, __resume and __abort] with gcode: PAUSE. RESUME and CANCEL_PRINT respectively.  PAUSE, RESUME and CANCEL_PRINT GCODE macros are defined in pause_resume.py and correctly save and restore GCODE state on pause and resume respectively.

After this change resume still didn't work from Repetier-Server and I'm not sure why because the RESUME macro should have been called to restore the state, I will look into this more. 

However when I changed pause_resume.py to call action:pause and action:resume, Repetier-Server called the GCODE macros and everything worked as expected.

I couldn't find anything in the Repetier docs, however, as it is very similar to OctoPrint, its docs stated (https://docs.octoprint.org/en/master/features/action_commands.html): "action:paused": When this command is received from the printer, OctoPrint will pause a current print job but without triggering any GCODE scripts or sending SD print control commands to the printer.
(which is the behaviour I was seeing)

@Arksine commented in PR #3857 indicated pause_resume used to use action:pause and resume, however it was changed to paused/resumed to cater for a bug in OctoPrint (which doesn't sound right to me, they should be fixing OctoPrint).

Long story short, we need a way to pause/resume the printer from Klipper, Host (OctoPrint/Repetier-Server/etc) and/or the Printer LCD, in a way that allows the user to reliably save the printer state on pause and restore it on resume.

**Comment by TheJoshW** [comment]
Ok did some instrumentation, and it appears that when resuming from the Host (OctoPrint or Repetier in my case), the RESUME GCODE is not being called when the printer is paused using action:paused.  May be a feature of Repetier, I'll check in their forums.

**Comment by TheJoshW** [comment]
Ok I have addressed this issue in the Repetier forums: https://forum.repetier.com/discussion/8608/resuming-from-repetier-server-after-pausing-from-action-paused.  The Repetier solution should fix this for Repetier hosts, and if OctoPrint still works then we can cancel this issue.  I will leave it open for now in case someone identified a more wholistic solution to this issue.

In retrospect, this actually highlights the power and flexibility of Klipper, allowing it to be controlled (and paused/resumed) from multiple places and still resume printing perfectly (if properly configured in the host as well).
